python/pyarrow/_dataset.pyx

```diff
@@ -783,11 +787,17 @@ cdef class FileFormat(_Weakrefable):
         if partition_expression is None:
             partition_expression = _true
 
-        c_source = _make_file_source(file, filesystem)
-        c_fragment = <shared_ptr[CFragment]> GetResultValue(
-            self.format.MakeFragment(move(c_source),
-                                     partition_expression.unwrap(),
-                                     <shared_ptr[CSchema]>nullptr))
+        cdef:
+            CFileSource c_source = _make_file_source(file, filesystem)
+            CResult[shared_ptr[CFileFragment]] c_result
+            CExpression c_partition_expression = partition_expression.unwrap()
+
+        with nogil:
+            c_result = self.format.MakeFragment(
```

Comment:
I wasn't able to write a test that needs this change for `make_fragment`. But I also suppose it doesn't hurt to add the `with nogil` ?

Reply:
If `MakeFragment` is self-contained and doesn't call into arbitrary IO code then it shouldn't be necessary.

Reply:
Indeed, it seems that `MakeFragment` just takes the input path / source and filesystem and puts that in a FileFragment, and for example doesn't actually check the file for the schema (that is only done the first time when accessing the schema in `ReadPhysicalSchema`, and that part in cython is correctly releasing the GIL)